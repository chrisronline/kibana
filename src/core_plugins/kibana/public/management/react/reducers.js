/* eslint-disable */
import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux'

import app from './app/app.reducer';
import indexPattern from './index-pattern/index-pattern.reducer';

export const rootReducer = combineReducers({
  routing: routerReducer,
  app,
  indexPattern,
});
