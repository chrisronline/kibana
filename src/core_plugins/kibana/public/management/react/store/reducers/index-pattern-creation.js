import { handleActions } from 'redux-actions';
import { createSelector } from 'reselect';
import { chunk, sortBy as sortByLodash } from 'lodash';

import {
  selectTimeField,
  fetchedTimeFields,
  goToNextPage,
  goToPreviousPage,
  changeSort,
  fetchedIndices,
  includeSystemIndices,
  excludeSystemIndices,
  creatingIndexPattern,
  createdIndexPattern,
} from '../actions/index-pattern-creation';

import {
  getIndexPatternCreate,
} from '../../reducers';

const defaultState = {
  isIncludingSystemIndices: false,
  isCreating: false,
  timeFields: {
    timeFields: undefined,
    selectedTimeField: undefined,
  },
  results: {
    indices: undefined,
    pattern: undefined,
    hasExactMatches: false,
    perPage: 10,
    page: 1,
    sortBy: undefined,
    sortAsc: true,
  }
}

export default handleActions({
  [selectTimeField](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [fetchedTimeFields](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [selectTimeField](state, { payload }) {
    const { timeField } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        selectedTimeField: timeField,
      },
    };
  },
  [fetchedIndices](state, { payload }) {
    const { indices, hasExactMatches, pattern } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        indices,
        hasExactMatches,
        pattern,
      },
    };
  },
  [changeSort](state, { payload }) {
    const { sortBy, sortAsc } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        sortBy,
        sortAsc: sortBy === state.results.sortBy ? !state.results.sortAsc: sortAsc,
      },
    };
  },
  [goToNextPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getNextPage(state.results),
      },
    };
  },
  [goToPreviousPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getPreviousPage(state.results),
      },
    };
  },
  [includeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: true,
    };
  },
  [excludeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: false,
    };
  },
  [creatingIndexPattern](state, action) {
    return {
      ...state,
      isCreating: true,
    };
  },
  [createdIndexPattern](state, action) {
    return {
      ...state,
      isCreating: false,
    };
  },
}, defaultState);

function getNextPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const nextPage = page + 1;
  return nextPage > pages
    ? 1
    : nextPage;
}

function getPreviousPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const previousPage = page - 1;
  return previousPage < 1
    ? pages
    : previousPage;
}

function getFilteredIndices(indices, isIncludingSystemIndices) {
  if (!indices) {
    return indices;
  }

  if (isIncludingSystemIndices) {
    return indices;
  }

  // All system indices begin with a period.
  return indices.filter(index => !index.name.startsWith('.'));
}

function getPaginatedIndices(indices, page, perPage) {
  const pagesOfIndices = chunk(indices, perPage);
  return pagesOfIndices[page - 1];
}

export const getFilteredAndPaginatedIndices = createSelector(
  [
    state => state.results,
    state => state.isIncludingSystemIndices
  ],
  ({
    indices: allIndices,
    page,
    perPage,
    sortBy,
    sortAsc,
  }, isIncludingSystemIndices) => {
    if (allIndices === undefined) {
      return {
        indices: undefined,
        numOfPages: 0,
      };
    }

    let filteredIndices = getFilteredIndices(allIndices, isIncludingSystemIndices);
    if (!!sortBy) {
      filteredIndices = sortByLodash(filteredIndices, sortBy);
      if (!sortAsc) {
        filteredIndices.reverse();
      }
    }

    const numOfPages = Math.ceil(filteredIndices.length / perPage);
    const indices = getPaginatedIndices(filteredIndices, page, perPage);
    return {
      indices,
      numOfPages,
    }
  }
);

export const getPattern = state => getIndexPatternCreate(state).results.pattern;
export const getSelectedTimeField = state => getIndexPatternCreate(state).timeFields.selectedTimeField;
