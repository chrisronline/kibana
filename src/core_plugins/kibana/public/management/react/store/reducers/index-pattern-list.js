import { handleActions } from 'redux-actions';

import {
  fetchedIndexPatterns,
  setTransientTableId,
} from '../actions/index-pattern-list';

import {
  getIndexPatternList,
} from '../../reducers';

const defaultState = {
  indexPatterns: undefined,
  transientTableId: undefined,
};

export default handleActions({
  [fetchedIndexPatterns](state, { payload }) {
    const { indexPatterns } = payload;

    return {
      ...state,
      indexPatterns,
    };
  },
  [setTransientTableId](state, { payload: { id: transientTableId } }) {
    return {
      ...state,
      transientTableId,
    };
  }
}, defaultState);

export const getIndexPatterns = state => getIndexPatternList(state).indexPatterns;
export const getPathToIndexPatterns = () => 'indexPatterns';
