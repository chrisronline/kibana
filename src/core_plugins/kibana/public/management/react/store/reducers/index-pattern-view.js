import { handleActions } from 'redux-actions';

import {
  fetchedIndexPattern,
  setAsDefaultIndexPattern,
} from '../actions/index-pattern-view';

import {
  getIndexPatternView,
} from '../../reducers';

const defaultState = {
  indexPattern: {
    pattern: undefined,
    fields: undefined,
    timeFieldName: undefined,
    isDefault: undefined,
    id: undefined,
  },
  fieldsTable: {},
  tabs: {},
};

export const indexPatternView = handleActions({
  [fetchedIndexPattern](state, { payload: { indexPattern } }) {
    return {
      ...state,
      indexPattern,
    };
  },
  [setAsDefaultIndexPattern](state) {
    return {
      ...state,
      indexPattern: {
        ...state.indexPattern,
        isDefault: true,
      }
    };
  }
}, defaultState);

export const getPathToFields = () => 'indexPattern.fields';
export const getTabs = state => getIndexPatternView(state).tabs;
