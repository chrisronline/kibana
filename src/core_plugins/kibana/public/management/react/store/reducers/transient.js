import { handleActions } from 'redux-actions';

import {
  change,
  createItem,
} from '../actions/transient';

const defaultState = {};

export default handleActions({
  [createItem](state, { payload: { id, defaultData }}) {
    return {
      ...state,
      [id]: defaultData || {},
    }
  },
  [change](state, { payload: { id, data } }) {
    return {
      ...state,
      [id]: {
        ...state[id],
        ...data
      },
    };
  },
}, defaultState);
