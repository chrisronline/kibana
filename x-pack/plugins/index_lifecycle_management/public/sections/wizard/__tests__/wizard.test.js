/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { Wizard } from '../';
import { indexLifecycleManagement } from '../../../store/reducers';

jest.mock('../../../api', () => ({
  bootstrap: jest.fn()
}));
jest.mock('../../../store/actions', () => ({
  saveLifecycle: jest.fn(),
}));
jest.mock('../components/index_template', () => ({
  IndexTemplate: () => '',
}));
jest.mock('../components/policy_configuration', () => ({
  PolicyConfiguration: () => '',
}));
jest.mock('../components/review', () => ({
  Review: () => '',
}));

jest.mock('@elastic/eui', () => ({
  EuiPage: props => props.children || '',
  EuiPageBody: props => props.children || '',
  EuiPageContent: props => props.children || '',
  EuiTitle: props => props.children || '',
  EuiSpacer: props => props.children || '',
  EuiStepsHorizontal: props => props.children || '',
}));

const defaultStore = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
};

describe('Wizard', () => {
  it('should render the first step', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <Wizard/>
      </Provider>
    );

    expect(component.find('EuiPageContent')).toMatchSnapshot();
  });

  it('should render the second step', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };
    const component = mount(
      <Provider store={store}>
        <Wizard/>
      </Provider>
    );

    component.find('Wizard').instance().setState({ selectedStep: 2 });
    component.update();

    expect(component.find('EuiPageContent')).toMatchSnapshot();
  });

  it('should render the third step', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };
    const component = mount(
      <Provider store={store}>
        <Wizard/>
      </Provider>
    );

    component.find('Wizard').instance().setState({ selectedStep: 3 });
    component.update();

    expect(component.find('EuiPageContent')).toMatchSnapshot();
  });

  it('should expose a addLifecycle method that saves the lifecycle and bootstraps the index/alias', async () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        },
        general: {
          ...defaultState.general,
          bootstrapEnabled: true,
          indexName: 'my_index',
          aliasName: 'my_alias',
        }
      }),
      ...defaultStore,
    };

    const actions = require('../../../store/actions');
    const api = require('../../../api');
    const component = mount(
      <Provider store={store}>
        <Wizard/>
      </Provider>
    );

    const lifecycle = { foo: 1 };
    await component.find('Wizard').instance().addLifecycle(lifecycle);
    expect(actions.saveLifecycle.mock.calls[0][0]).toEqual(lifecycle);
    expect(actions.saveLifecycle.mock.calls[0][1]).toEqual({
      indexTemplate: '',
      primaryShardCount: 1,
      replicaCount: 1,
      lifecycleName: '',
      nodeAttrs: ''
    });
    expect(api.bootstrap).toHaveBeenCalledWith('my_index', 'my_alias');
  });
});
