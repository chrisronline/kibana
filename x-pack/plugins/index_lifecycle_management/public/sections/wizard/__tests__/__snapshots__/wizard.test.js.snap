// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard should render the first step 1`] = `
<EuiPageContent
  className="ilmContent"
  horizontalPosition="center"
  verticalPosition="center"
>
  <EuiTitle
    size="m"
  >
    <h2>
      Index lifecycle management
    </h2>
  </EuiTitle>
  <EuiSpacer />
  <EuiStepsHorizontal
    steps={
      Array [
        Object {
          "isComplete": false,
          "isSelected": true,
          "onClick": [Function],
          "title": "Select a template",
        },
        Object {
          "disabled": true,
          "isComplete": false,
          "isSelected": false,
          "onClick": [Function],
          "title": "Configure a policy",
        },
        Object {
          "disabled": true,
          "isComplete": false,
          "isSelected": false,
          "onClick": [Function],
          "title": "Review and save",
        },
      ]
    }
  />
  <EuiSpacer />
  <IndexTemplate
    done={[Function]}
    errors={
      Object {
        "configuration": Object {
          "node_attrs": Array [],
          "primary_nodes": Array [],
          "replicas": Array [],
        },
        "templateSelection": Object {
          "aliasName": Array [],
          "indexName": Array [],
          "templateName": Array [
            "An index template is required",
          ],
        },
      }
    }
    validate={[Function]}
  />
</EuiPageContent>
`;

exports[`Wizard should render the second step 1`] = `
<EuiPageContent
  className="ilmContent"
  horizontalPosition="center"
  verticalPosition="center"
>
  <EuiTitle
    size="m"
  >
    <h2>
      Index lifecycle management
    </h2>
  </EuiTitle>
  <EuiSpacer />
  <EuiStepsHorizontal
    steps={
      Array [
        Object {
          "isComplete": true,
          "isSelected": false,
          "onClick": [Function],
          "title": "Select a template",
        },
        Object {
          "disabled": false,
          "isComplete": false,
          "isSelected": true,
          "onClick": [Function],
          "title": "Configure a policy",
        },
        Object {
          "disabled": true,
          "isComplete": false,
          "isSelected": false,
          "onClick": [Function],
          "title": "Review and save",
        },
      ]
    }
  />
  <EuiSpacer />
  <PolicyConfiguration
    back={[Function]}
    done={[Function]}
    errors={
      Object {
        "cold": Object {
          "selectedAfter": Array [],
          "selectedAfterUnits": Array [],
          "selectedAlias": Array [],
          "selectedNodeAttrs": Array [],
          "selectedReplicaCount": Array [],
        },
        "delete": Object {
          "selectedAfter": Array [],
          "selectedAfterUnits": Array [],
          "selectedAlias": Array [],
        },
        "hot": Object {
          "selectedAlias": Array [],
          "selectedMaxAge": Array [
            "A maximum age is required",
          ],
          "selectedMaxAgeUnits": Array [],
          "selectedMaxDocSize": Array [],
          "selectedMaxSizeStored": Array [
            "A maximum index size is required",
          ],
          "selectedMaxSizeStoredUnits": Array [],
        },
        "warm": Object {
          "selectedAfter": Array [],
          "selectedAfterUnits": Array [],
          "selectedAlias": Array [],
          "selectedForceMergeSegments": Array [],
          "selectedNodeAttrs": Array [],
          "selectedPrimaryShardCount": Array [],
          "selectedReplicaCount": Array [],
        },
      }
    }
    validate={[Function]}
  />
</EuiPageContent>
`;

exports[`Wizard should render the third step 1`] = `
<EuiPageContent
  className="ilmContent"
  horizontalPosition="center"
  verticalPosition="center"
>
  <EuiTitle
    size="m"
  >
    <h2>
      Index lifecycle management
    </h2>
  </EuiTitle>
  <EuiSpacer />
  <EuiStepsHorizontal
    steps={
      Array [
        Object {
          "isComplete": true,
          "isSelected": false,
          "onClick": [Function],
          "title": "Select a template",
        },
        Object {
          "disabled": false,
          "isComplete": true,
          "isSelected": false,
          "onClick": [Function],
          "title": "Configure a policy",
        },
        Object {
          "disabled": false,
          "isComplete": false,
          "isSelected": true,
          "onClick": [Function],
          "title": "Review and save",
        },
      ]
    }
  />
  <EuiSpacer />
  <Review
    back={[Function]}
    done={[Function]}
    errors={
      Object {
        "policyName": Array [
          "A policy name is required",
        ],
      }
    }
    validate={[Function]}
  />
</EuiPageContent>
`;
